### ISSUE TITLE
Discuss: What if a user just told us if they wanted a non-standard save directory?

### ISSUE THREAD

**Issue by resinneublem** [pyfa-org/Pyfa#908]
@blitzmann I came across https://www.reddit.com/r/Eve/comments/5iirjh/how_to_keep_pyfa_synced_between_multiple_machines/ today which was funny because I was talking with a corpmate about https://github.com/pyfa-org/Pyfa/wiki/Cross-Platform-Data-Sharing yesterday.

> i'm cool with that if you make 'symlink/junction' something a non-developer can understand

What if we simply allowed the user to change the directory where the .pyfa directory is stored within the GUI somewhere in the preferences? It would probably require a restart, but allowing people just to edit their "workspace" seems like it'd be a pretty big usability win. They wouldn't have to muck around with shortcuts or junctions/symlinks.

`-s /path/to/directory` would still take the highest precedence.

**Comment by Ebag333** [comment]
You can already change it to be in the install dir.

Easy solution, put on Google Drive (or Amazon, or Box, or whatever), change command line to save user data in install dir, done.

**Comment by resinneublem** [comment]
So if you wanted from a normal installation and wanted to use dropbox you'd have to uninstall and reinstall? And then of course move your directory manually to preserve all your existing fits?

That's a big headache that a single setting can avoid.

**Comment by Ebag333** [comment]
A setting won't move the user DB.  I mean, sure, it _could_ if we programmed it to, but that's a lot more work.  And there's no reason we couldn't do that for the built in flags.

Here's how it'd work.

Initial setup:
1. Download and install Pyfa to your cloud drive.
2. Change shortcut to run with the flag to save user data locally.
3. Done.

New computer:
1. Install cloud drive software.
2. Run Pyfa.
3. Done.

There's no uninstall and reinstall, so not sure what you mean by that.  If a user is converting from an existing install, they can just move the relevant files over.

I would like to eventually handle moving the files around for the user, and make the database save path modifiable from within Pyfa (as well as persistence, without having to use external flags).

This is the first step in that direction: #868

**Comment by resinneublem** [comment]
> A setting won't move the user DB. I mean, sure, it could if we programmed it to, but that's a lot more work. 

That is precisely why I started a **discussion** thread to talk about this.

> Here's how it'd work.

You're talking about the way things currently work and I'd like to discuss something new. Not to mention there's TONS of ways to implement this. I'd first like to discuss the idea and then move on to different implementations instead of just assuming how it's going to work.

> There's no uninstall and reinstall, so not sure what you mean by that. If a user is converting from an existing install, they can just move the relevant files over.

With the current setup and code there absolutely is if they have already installed pyfa to the default location. Otherwise they have to modify shortcuts like that reddit post mentions.

I started this discussion thread to talk saving user work on having to mess with shortcuts, junctions and symlinks.

> I would like to eventually handle moving the files around for the user, and make the database save path modifiable from within Pyfa (as well as persistence, without having to use external flags).

Dude. I don't understand why you are so argumentative and possessive about the way Pyfa should or should not be and then turn around and restate EXACTLY what I'm proposing we talk about here.

**Comment by Ebag333** [comment]
>Dude. I don't understand why you are so argumentative and possessive about the way Pyfa should or should not be and then turn around and restate EXACTLY what I'm proposing we talk about here.

I'm not trying to be argumentative or possessive, what you started the thread was not exactly what it seems you want to discuss.

From the original post:
>What if we simply allowed the user to change the directory where the .pyfa directory is stored within the GUI somewhere in the preferences? It would probably require a restart, but allowing people just to edit their "workspace" seems like it'd be a pretty big usability win. 

We already do allow the user to change the location (though to be fair we only allow one of two places).  

No where does it talk about moving database/settings/etc, which is a lot more complex.  If you want a discussion, you shouldn't assume someone is trying to argue with you just because they brought up something you didn't. :)

>They wouldn't have to muck around with shortcuts or junctions/symlinks.

And my second post is addressing this bit.  There's no need to mess with junctions/symlinks, that's most definitively not the easy way to do it.

>`I would like to`

This doesn't mean `so it has been said, so shall it be done`.  It means just what it says, I _want_ to do this but that doesn't mean that it'll happen.  I actually started down the path of making the DB save path modifiable from inside Pyfa, but @blitzmann wasn't really on board with it at the time, largely (I believe, it was a while ago so might be mis-remembering) due to the same things I basically mentioned above.  So it got scaled back to just exposing it, to make it easier for end users to find it.

If you want to take that and code up something to handle moving the DB, preferences, logs, etc, more power to you.  Go for it, I'm not opposed to it at all, quite the opposite.

Just a word of warning, the problem I ran into is that one of the first things Pyfa does is lock up the files it uses.  So you'll have to figure out a way to trigger it from inside Pyfa, but handle it outside of the main thread. Or perhaps flag a setting so the next Pyfa restart it'll know to move it (before it grabs the files).

**Comment by resinneublem** [comment]
> If you want to take that and code up something to handle moving the DB, preferences, logs, etc, more power to you. Go for it, I'm not opposed to it at all, quite the opposite.

Awesome. I took from all of that...you're on board with this idea. That was a ton of words to get to that point.

I'll wait until @blitzmann weighs in on this idea before I dive in to any code, screenshots or implementation discussions.

**Comment by blitzmann** [comment]
So, the problem here lies with two issues:

1) All settings are currently saved in the saved user path. So, it stands to reason there is a bit of a logistical problem there. You start in `~/.pyfa`, and in the settings you choose to save in `~/Desktop/EVE/pyfa`. However, that settings is stored in the old directory... so, how do we switch over to the one the user chose? Right now there isn't a central settings directory that persists very well. Obviously, it doesn't have to be this way, but any solution probably would require moving around files, which brings me to...

2) Moving around user files generally makes me nervous. You fuck up a condition, and now the save path is all wrong and not pointing to anything useful and the user starts with (presumably) a fresh DB, or worse pyfa crashes at the get go. Or even worse: the database is deleted / corrupted / lost and otherwise unrecoverable, and the user starts with a fresh DB / pyfa crashes.

These are not unsolvable problems, but they are problems that require a bit more engineering than I was willing to put in, especially due to workarounds being available. 

To do this, we will probably need to require a settings files that -does not change- regardless of what the user chooses. A central config file that basically points pyfa to the other areas it needs to be, one that is stored in a writable area for the user. Or hell, maybe all settings stick around in `~/.pyfa` and we only move the database around. 

There's many ways to do this. If you're willing to take it on, I'm willing to listen to any proposal you may have on how to implement it. :)

Also, this might be a thing you might want to look at #237

**Comment by resinneublem** [comment]
> To do this, we will probably need to require a settings files that -does not change- regardless of what the user chooses. A central config file that basically points pyfa to the other areas it needs to be, one that is stored in a writable area for the user. 

Yeah, that's where my brain was headed too. This is what was going on in my mind: having `~/.pyfaconfig` or something that will point to a "fitting database" or something. (Terminology and file names aren't important just yet. I'm just throwing those names out as an example). I'm thinking a bit of how Eclipse handles different workspaces.

A few things that would go into this:

1. If a user changes the path to a new area, there could be an additional checkbox for Pyfa to copy the existing database over. We wouldn't ever move things or delete the old stuff. That also avoids logistical file locking issues with open files.
2. This would probably require the user restarting Pyfa. Avoids any sort of hotswapping of the database and reconnecting.
3. _Maybe_ also have something in the UI that tells them they have a database with fits in `~/.pyfa` and/or a very easy way to reset to the default without actually blasting the data. It gives them the flexibility of trying out a new feature and reverting it if they don't like it without a lot of risk.
4. _Maybe_ would be nice to have a list somewhere of the last known "fitting databases". I'm kinda thinking of "Open Recent File..." in most editors/apps.

https://github.com/pyfa-org/Pyfa/issues/237 is super interesting though because it kinda makes me think that even `~/.pyfaconfig` (or some other file) in the home directory isn't appropriate. I think this would take some thinking/tinkering. Perhaps doing the "right thing" by storing the default config and database in the a standard location on Windows (Documents), macOS (My Documents), Linux (`.config/` or whatever is appropriate there. Like `~/.local/...`).

**However. I don't think the introduction of this feature should "migrate" anyone. I think this could be rolled out to grandfather existing users and leave their database in the existing directory. It might just create the config file on boot to point to that directory.**

My goal for this would be for the user to deviate very easily from `~/.pyfa` from within the application in case they want to store it anywhere else, like a Dropbox folder, or a thumbdrive.

**Comment by Ebag333** [comment]
I really like the idea of being able to open databases.  That'd allow _much_ easier sharing of databases than what we have currently.

>#237 is super interesting though because it kinda makes me think that even ~/.pyfaconfig (or some other file) in the home directory isn't appropriate. I think this would take some thinking/tinkering. Perhaps doing the "right thing" by storing the default config and database in the a standard location on Windows (Documents), macOS (My Documents), Linux (.config/ or whatever is appropriate there. Like ~/.local/...).

From a Windows perspective, there's problems with our name and location.  `.pyfa` isn't really acceptable because directories aren't supposed to start with special characters.  The location is also problematic (locked down users don't have rights there).

So it should default to: `C:\Users\User\AppData\Local\pyfa`

If we want to put settings there, that's probably where it should go as well.

I wouldn't use Documents for settings and config, that's not really standard practice for Windows.  The database can go there, but users often don't like things cluttering up their documents folder (so give them the option to put it there, but default it to AppData\Local).

**Comment by resinneublem** [comment]
> So it should default to: C:\Users\User\AppData\Local\pyfa

Yeah now that you mention it, that makes a lot more sense.
